File: 043_multiply_strings/Solution1/Solution.go
package main

import (
    "fmt"
    "strconv"
    "strings"
)

func multiply(num1 string, num2 string) string {
    var result string
    n1, n2 := len(num1) - 1, len(num2) - 1
    if num1 == "0" || num2 == "0" {
        return "0"
    }
    carry := 0

    for j := n2; j >= 0; j-- {
        curNum2 := getNum(num2, j)
        var tmpRes []string
        for i := n1; i >= 0 || carry != 0; i-- {
            curNum1 := getNum(num1, i)
            tmp := curNum1 * curNum2 + carry
            carry = tmp / 10
            tmpRes = append(tmpRes, strconv.Itoa(tmp % 10))
        }
        result = addStrings(result, strings.Join(reverseArr(tmpRes), ""), n2 - j)
    }

    return result
}

func addStrings(s1, s2 string, digitCount int) string {
    for i := 0; i < digitCount; i++ {
        s2 += "0"
    }
    n1, n2 := len(s1) - 1, len(s2) - 1;
    var result []string
    carry := 0

    for n1 >= 0 || n2 >= 0 || carry != 0 {
        x := getNum(s1, n1)
        y := getNum(s2, n2)
        tmp := x + y + carry
        carry = tmp / 10
        result = append(result, strconv.Itoa(tmp % 10))
        n1--
        n2--
    }

    return strings.Join(reverseArr(result), "")
}

func reverseArr(s []string) []string {
    for i, j := 0, len(s) - 1; i < j; i, j = i + 1, j - 1 {
        s[i], s[j] = s[j], s[i]
    }

    return s
}

func getNum(num string, index int) int {
    if index >= 0 {
        return int(num[index] - '0')
    } else {
        return 0
    }
}

func main() {
    fmt.Println(multiply("123", "45"))
}

File: 043_multiply_strings/Solution1/Solution.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        result = ''
        if num1 == '0' or num2 == '0': return '0'
        n1, n2 = len(num1) - 1, len(num2) - 1
        carry = 0

        for j in range(n2, -1, -1):
            cur_num2 = int(num2[j])
            tmp_result = []
            i = n1
            while i >= 0 or carry != 0:
                cur_num1 = int(num1[i]) if i >= 0 else 0
                tmp = cur_num1 * cur_num2 + carry
                carry = tmp // 10
                tmp_result.append(str(tmp % 10))
                i -= 1
            result = self.add_strings(result, ''.join(reversed(tmp_result)), n2 - j)

        return result

    def add_strings(self, s1: str, s2: str, digit_count: int) -> str:
        s2 = s2 + ''.join(['0' for i in range(digit_count)])
        n1, n2 = len(s1) - 1, len(s2) - 1
        carry = 0
        result = []

        while n1 >= 0 or n2 >= 0 or carry != 0:
            x = int(s1[n1]) if n1 >= 0 else 0
            y = int(s2[n2]) if n2 >= 0 else 0
            tmp = x + y + carry
            carry = tmp // 10
            result.append(str(tmp % 10))
            n1 -= 1
            n2 -= 1

        return ''.join(reversed(result))


if __name__ == '__main__':
    solution = Solution()
    print(solution.multiply('123', '45'))

